### PR TITLE
EmitC: Add emitc.global and emitc.get_global (#145)

### PR DESCRIPTION
**Pull request by mgehre-amd**
This adds
- `emitc.global` and `emitc.get_global` ops to model global variables similar to how `memref.global` and `memref.get_global` work.
- translation of those ops to C++
- lowering of `memref.global` and `memref.get_global` into those ops

### CODE REVIEW

**Comment by simon-camp** [comment]
I've got a high level question before reviewing thoroughly.

Should the `constness` of globals (and values in general) be modeled at the type level? Currently it wouldn't be possible to take the address of a const global for example.

**Comment by mgehre-amd** [comment]
> I've got a high level question before reviewing thoroughly.
> 
> Should the `constness` of globals (and values in general) be modeled at the type level? Currently it wouldn't be possible to take the address of a const global for example.

I view emitc as a mostly syntactic dialect, so `const` needs to be annotated at all places where we need to write `const` in the C++ code. I don't think the type system will need to carry it.

What do you mean by `it wouldn't be possible to take the address of a const global`? I'd think you can call emitc.apply on the result of a `emitc.get_global` to get the address of that global, independent of whether it is const.

**Review by simon-camp — CHANGES_REQUESTED** [review]
I have two two questions/suggestions. Otherwise this looks great and adds one of the big missing features, thanks for working on this.

**Comment by simon-camp** [comment]
> What do you mean by `it wouldn't be possible to take the address of a const global`? I'd think you can call emitc.apply on the result of a `emitc.get_global` to get the address of that global, independent of whether it is const.

You are right, I thought we were verifying operand type and pointee type in the `apply &` op. We would need a way to represent `pointer to const int` for example, but let's postpone this discussion to a later point if/when need for this comes up.

File: mlir/lib/Dialect/EmitC/IR/EmitC.cpp [hunk 1]
```diff
@@ -896,3 +889,111 @@ LogicalResult mlir::emitc::OpaqueType::verify(
...
+        return emitOpError("initial value expected to be of type ")
+               << getType() << ", but was of type " << intAttr.getType();
+      }
+    } else if (auto floatAttr = dyn_cast<FloatAttr>(initValue)) {
+      if (floatAttr.getType() != getType()) {
+        return emitOpError("initial value expected to be of type ")
+               << getType() << ", but was of type " << floatAttr.getType();
+      }
+    } else {
+      return emitOpError(
```
[the remark this answers is not in the document]
Comment (mgehre-amd) [reply mlir/lib/Dialect/EmitC/IR/EmitC.cpp:961]: True, I now allow opaque attributes here.

**Review by marbre — COMMENTED** [review]
Nice to see globals support rising! Some minor remarks after going though part of the newly added code.

File: mlir/lib/Target/Cpp/TranslateToCpp.cpp [hunk 2]
```diff
@@ -1384,6 +1403,30 @@ LogicalResult CppEmitter::emitVariableDeclaration(OpResult result,
...
+  if (op.getConstSpecifier())
+    os << "const ";
+
+  if (failed(emitVariableDeclaration(op->getLoc(), op.getType(),
+                                     op.getSymName()))) {
+    return failure();
+  }
+
+  std::optional<Attribute> initialValue = op.getInitialValue();
+  if (initialValue && !isa<UnitAttr>(*initialValue)) {
```
Comment (simon-camp) [inline mlir/lib/Target/Cpp/TranslateToCpp.cpp:1420]:
```suggestion
  if (initialValue) {
```

File: mlir/test/Conversion/MemRefToEmitC/memref-to-emitc.mlir [hunk 3]
```diff
@@ -26,3 +27,19 @@ func.func @memref_load(%i: index, %j: index) -> f32 {
   // CHECK: return %[[VAR]] : f32
   return %1 : f32
 }
+
+// -----
+
+// CHECK-LABEL: globals
+module @globals {
+  memref.global "private" constant @internal_global : memref<3x7xf32> = dense<4.0>
```
Comment (simon-camp) [inline mlir/test/Conversion/MemRefToEmitC/memref-to-emitc.mlir:35]: We should have a test for uninitialized globals like
```mlir
memref.global @uninit : memref<2xf32> = uninitialized
```

File: mlir/lib/Conversion/MemRefToEmitC/MemRefToEmitC.cpp [hunk 4]
```diff
@@ -50,6 +50,68 @@ struct ConvertAlloca final : public OpConversionPattern<memref::AllocaOp> {
...
+          op.getLoc(),
+          "only public and private visibility is currently supported");
+    }
+    // We are explicit in specifier the linkage because the default linkage
+    // for constants is different in C and C++.
+    bool staticSpecifier = visibility == SymbolTable::Visibility::Private;
+    bool externSpecifier = !staticSpecifier;
+
+    rewriter.replaceOpWithNewOp<emitc::GlobalOp>(
+        op, operands.getSymName(), resultTy, operands.getInitialValueAttr(),
```
Comment (simon-camp) [inline mlir/lib/Conversion/MemRefToEmitC/MemRefToEmitC.cpp:90]: We need to explicitly handle uninitialized globals. (By mapping UnitAttr to std::nullopt I think)

**Review by simon-camp — CHANGES_REQUESTED** [review]

File: mlir/lib/Conversion/MemRefToEmitC/MemRefToEmitC.cpp (hunk 4 above)
Comment (mgehre-amd) [reply mlir/lib/Conversion/MemRefToEmitC/MemRefToEmitC.cpp:90]: done

**Review by simon-camp — APPROVED** [review]
LGTM, thanks.

**Pull request merged**